gbuiltins/staking/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+//! for any Gear runtime and provide an API for the applications to build on top
+//! of some blockchain logic like staking, governance, etc.
+
+//! For a builtin actor to process a message, it should be able to decode its
+//! payload into one of the supported message types.
+//!
+//! # Examples
+//!
+//! The following example shows how a contract can send a message to a builtin actor
+//! (specifically, a staking actor) to bond some `value` to self as the controller
+//! so that the contract can later use the staking API to nominate validators.
+//!
+//! ```ignore
+//! use gstd::{msg, ActorId};
+//! use gbuiltins::staking::{Request, RequestV1, RewardAccount};
+//! use parity_scale_codec::Encode;
+//!
+//! const BUILTIN_ADDRESS: ActorId = ActorId::new(hex_literal::hex!(
+//!     "77f65ef190e11bfecb8fc8970fd3749e94bed66a23ec2f7a3623e785d0816761"
+//! ));
+//!
+//! #[gstd::async_main]
+//! async fn main() {
+//!     let value = msg::value();
+//!     let payee: Option<RewardAccount> = None;
+//!     let payload = Request::V1(RequestV1::Bond { value, payee }).encode();
+//!     let _ = msg::send_bytes_for_reply(BUILTIN_ADDRESS, &payload[..], 0, 0)
+//!         .expect("Error sending message")
+//!         .await;
+//! }
+//! # fn main() {}
+//! ```
+//!
+
+#![no_std]
+
+extern crate alloc;
+
+use alloc::vec::Vec;
+use scale_info::scale::{self, Decode, Encode};
+
+pub type AccountId = [u8; 32];
+
+/// Type that should be used to create a message to the staking built-in actor.
+///
+/// A [partial] mirror of the staking pallet interface. Not all extrinsics
+/// are supported, more can be added as needed for real-world use cases.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+#[codec(crate = scale)]
+pub enum Request {
+    /// Version 1 of the staking built-in actor protocol.
+    V1(RequestV1),
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+#[codec(crate = scale)]
+pub enum RequestV1 {
+    /// Bond up to the `value` from the sender to self as the controller.
+    Bond {
+        value: u128,
+        payee: Option<RewardAccount>,
+    },
+    /// Add up to the `value` to the sender's bonded amount.
+    BondExtra { value: u128 },
+    /// Unbond up to the `value` to allow withdrawal after undonding period.
+    Unbond { value: u128 },
+    /// Withdraw unbonded chunks for which undonding period has elapsed.
+    WithdrawUnbonded { num_slashing_spans: u32 },
+    /// Add sender as a nominator of `targets` or update the existing targets.
+    Nominate { targets: Vec<AccountId> },
+    /// Declare intention to [temporarily] stop nominating while still having funds bonded.
+    Chill,
+    /// Request stakers payout for the given era.
+    PayoutStakers {
+        validator_stash: AccountId,
+        era: u32,
+    },
+    /// Rebond a portion of the sender's stash scheduled to be unlocked.
+    Rebond { value: u128 },
+    /// Set the reward destination.
+    SetPayee { payee: RewardAccount },
```

Comment:
it could be Option if you really wish but actually can't see any reason not to accept just address here

Reply:
Are we talking about having a dedicated enum `RewardAccount` indicating the reward destination here?
Generally, it is supposed to be a mirror type for the [RewardDestination<AccountID>](https://github.com/paritytech/polkadot-sdk/blob/f13408d57f7ac2c1f49cc9a2f95393c0ad4b9a35/substrate/frame/staking/src/lib.rs#L401) which is more than a wrapper around an address.
Initially, I only kept two variants of that one (thinking there wouldn't be much use for the remaining 3 in our current context) but now I'm more convinced that even with the `Controller` variant having been deprecated we still should stick to one-to-one correspondence with the target type to support the whole variety of possible use cases that may arise in applications later.